File: src/abrain/_cpp/phenotype/ann.cpp
#include <memory>
#include <cassert>
#include <algorithm>

#include "../config.h"
#include "../misc/utils.hpp"
#include "ann.h"

#include <iostream>

namespace kgd::eshn::phenotype {

#ifndef NDEBUG
//#define DEBUG_COMPUTE 1
#endif

#ifndef NDEBUG  // Timing utilities
using timing_clock = std::chrono::high_resolution_clock;
using time_point = std::chrono::time_point<timing_clock>;

static time_point t_now() { return timing_clock::now(); }

template <typename ANN>
static typename ANN::Stats::rep t_diff(const time_point start) {
  return std::chrono::duration_cast<typename ANN::Stats::duration>(t_now() - start).count();
}
#endif

template <unsigned int DI>
bool ANN_t<DI>::empty(bool strict) const {
  return strict ? (stats().hidden == 0) : (stats().edges == 0);
}

template <unsigned int DI>
bool ANN_t<DI>::perceptron() const {
  return (stats().hidden == 0) && stats().edges > 0;
}

template <unsigned int DI>
ANN_t<DI> ANN_t<DI>::build (
    const Coordinates &inputs,
    const Coordinates &outputs,
    const genotype::CPPNData &genome) {

  if (genome.inputs < 2*DI + genome.bias || 2*DI + genome.bias + 1 < genome.inputs)
    throw std::invalid_argument(utils::mergeToString(
      "Unable to build a ", DI, "D ANN from a genome with ",
      genome.inputs, " inputs. Did you mix it up with the one for the body?"
    ));
  using ANN = ANN_t<DI>;

#ifndef NDEBUG
const auto start_time = timing_clock::now();
#endif

  static const auto& weightRange = Config::annWeightsRange;

  CPPN cppn (genome);

  ANN ann;

  NeuronsMap &neurons = ann._neurons;

  const auto add = [&cppn, &ann] (auto p, auto t) {
    float bias = 0;
    if (t != Neuron::I)
      bias = cppn(p, Point::null(), CPPN::Output::BIAS);
    return ann.addNeuron(p, t, bias);
  };

  Coordinates sortedInputs (inputs);
  std::ranges::sort(sortedInputs.begin(), sortedInputs.end(), NeuronCMP());

  Coordinates sortedOutputs (outputs);
  std::ranges::sort(sortedOutputs.begin(), sortedOutputs.end(), NeuronCMP());

  unsigned int i = 0;
  ann._inputs.resize(inputs.size());
  ann._ibuffer.resize(inputs.size());
  for (auto &p: sortedInputs) neurons.insert(ann._inputs[i++] = add(p, Neuron::I));

  i = 0;
  ann._outputs.resize(outputs.size());
  ann._obuffer.resize(outputs.size());
  for (auto &p: sortedOutputs) neurons.insert(ann._outputs[i++] = add(p, Neuron::O));

  Coordinates hidden;
  evolvable_substrate::Connections_t<DI> connections;
  if (evolvable_substrate::connect(cppn, sortedInputs, sortedOutputs,
                                   hidden, connections,
                                   ann._stats.iterations)) {
    for (auto &p: hidden) neurons.insert(add(p, Neuron::H));
    for (auto &c: connections)
      ann.neuronAt(c.to)->addLink(c.weight * weightRange, ann.neuronAt(c.from));
  }

  ann.computeStats();

#ifndef NDEBUG
  ann._stats.time.build = t_diff<ANN>(start_time);
  ann._stats.time.eval = typename ANN::Stats::rep{0};
#endif

  return ann;
}

// Deepcopy implementation
// Not used for now
//void ANN::copyInto(ANN &that) const {
//  // Copy neurons
//  for (const Neuron::ptr &n: _neurons) {
//    Neuron::ptr n_ = that.addNeuron(n->pos, n->type, n->bias);
//    that._neurons.insert(n_);
//    n_->value = n->value;
//    n_->depth = n->depth;
//    n_->flags = n->flags;
//  }

//  // Generates links
//  for (const Neuron::ptr &n: _neurons) {
//    const Neuron::ptr n_ = that.neuronAt(n->pos);
//    for (const Neuron::Link &l: n->links()) {
//      n_->addLink(l.weight, that.neuronAt(l.in.lock()->pos));
//    }
//  }

//  // Update I/O buffers
//  that._inputs.reserve(_inputs.size());
//  for (const Neuron::ptr &n: _inputs)
//    that._inputs.push_back(that.neuronAt(n->pos));

//  that._outputs.reserve(_outputs.size());
//  for (const Neuron::ptr &n: _outputs)
//    that._outputs.push_back(that.neuronAt(n->pos));

//  // Copy stats
//  that._stats = _stats;
//}

template <unsigned int DI>
void ANN_t<DI>::reset() {
  for (auto &p: _neurons)   p->reset();
}

template <unsigned int DI>
unsigned int ANN_t<DI>::max_hidden_neurons() {
  return std::pow(2, DI*Config::maxDepth);
}

template <typename ANN>
unsigned int computeDepth (ANN &ann) {
  using Neuron = typename ANN::Neuron;
  using NeuronPtr = typename ANN::NeuronPtr;
  using Point = typename ANN::Point;
  struct ReverseNeuron {
    Neuron &n;
    std::vector<ReverseNeuron*> o;
    explicit ReverseNeuron (Neuron &n) : n(n) {}
  };

  std::map<Point, ReverseNeuron*> neurons;
  std::set<ReverseNeuron*> next;

  unsigned int hcount = 0, hseen = 0;
  for (const NeuronPtr &n: ann.neurons()) {
    auto p = neurons.emplace(std::make_pair(n->pos, new ReverseNeuron(*n)));
    if (n->type == Neuron::I) next.insert(p.first->second);
    else if (n->type == Neuron::H) hcount++;
  }

  for (const NeuronPtr &n: ann.neurons())
    for (auto &l: n->links())
      neurons.at(l.in.lock()->pos)->o.push_back(neurons.at(n->pos));

  std::set<ReverseNeuron*> seen;
  unsigned int depth = 0;
  while (!next.empty()) {
    auto current = next;
    next.clear();

    for (ReverseNeuron *n: current) {
      n->n.depth = depth;
//      std::cerr << n->n.pos << ": " << depth << "\n";
      seen.insert(n);
      if (n->n.type == Neuron::H)  hseen++;
      for (ReverseNeuron *o: n->o) next.insert(o);
    }

    decltype(seen) news;
    std::set_difference(next.begin(), next.end(),
                        seen.begin(), seen.end(),
                        std::inserter(news, news.end()));
    next = news;
    assert(hseen == hcount || next.size() > 0);

    depth++;
  }

  unsigned int d = 0;
  for (auto &p: neurons) {
    d = std::max(d, p.second->n.depth);
    delete p.second;
  }
  return d;
}

template <unsigned int DI>
void ANN_t<DI>::computeStats() {
  using Link = typename Neuron::Link;
  auto &e = _stats.edges = 0;
  float &l = _stats.axons = 0;

  float &u = _stats.utility = 0;
  std::set<Point> connected_inputs;

  for (const NeuronPtr &n: _neurons) {
    e += static_cast<unsigned int>(n->links().size());
    for (const Link &link: n->links()) {
      const auto &_n = link.in.lock();
      l += (n->pos - _n->pos).length();

      if (_n->type == Neuron::I)
        connected_inputs.insert(_n->pos);
    }

    if (n->type == Neuron::O && n->links().size() > 0) u++;
  }

  u += connected_inputs.size();
  u /= float(_inputs.size() + _outputs.size());

  _stats.hidden = static_cast<unsigned int>(_neurons.size() - _inputs.size() - _outputs.size());
  _stats.density = _stats.edges / float(max_edges());

  if (_stats.hidden == 0) {
    for (NeuronPtr &n: _inputs)   n->depth = 0;
    for (NeuronPtr &n: _outputs)  n->depth = 1;
    _stats.depth = 1;

  } else
    _stats.depth = computeDepth(*this);
}

template <unsigned int DI>
typename ANN_t<DI>::NeuronPtr
ANN_t<DI>::addNeuron(const Point &p, typename Neuron::Type t, float bias) {
  return std::make_shared<Neuron>(p, t, bias);
}

template <unsigned int DI>
void ANN_t<DI>::operator() (const IBuffer &inputs, OBuffer &outputs, unsigned int substeps) {
#ifndef NDEBUG
  const auto start_time = t_now();
#endif

  static const auto &activation =
    phenotype::CPPN::functions.at(Config::activationFunc);
  assert(inputs.size() == _inputs.size());
  assert(outputs.size() == outputs.size());

  for (unsigned int i=0; i<inputs.size(); i++) _inputs[i]->value = inputs[i];

#ifdef DEBUG_COMPUTE
  std::cerr << std::setprecision(std::numeric_limits<float>::max_digits10);
  using utils::operator<<;
  std::cerr << "## Compute step --\n inputs:\t" << inputs << "\n";
#endif

  for (unsigned int s = 0; s < substeps; s++) {
#ifdef DEBUG_COMPUTE
    std::cerr << "#### Substep " << s+1 << " / " << substeps << "\n";
#endif

    for (const auto &p: _neurons) {
      if (p->isInput()) continue;

      float v = p->bias;
      for (const auto &l: p->links()) {
#if DEBUG_COMPUTE >= 3
        std::cerr << "        i> v = " << v + l.weight * l.in.lock()->value
                  << " = " << v << " + " << l.weight << " * "
                  << l.in.lock()->value << "\n";
#endif

        v += l.weight * l.in.lock()->value;
      }

      p->value = activation(v);
      assert(-1 <= p->value && p->value <= 1);

#if DEBUG_COMPUTE >= 2
      std::cerr << "      <o " << p->pos << ": " << p->value << " = "
                << config::EvolvableSubstrate::activationFunc() << "("
                << v << ")\n";
#endif
    }
  }

  for (unsigned int i=0; i<_outputs.size(); i++)  outputs[i] = _outputs[i]->value;

#ifdef DEBUG_COMPUTE
  std::cerr << "outputs:\t" << outputs << "\n## --\n";
#endif

#ifndef NDEBUG
  _stats.time.eval += t_diff<ANN_t<DI>>(start_time);
#endif

}

template class ANN_t<2>;
template class ANN_t<3>;

} // end of namespace kgd::eshn::genotype

File: src/abrain/_python/ann.py
import dataclasses
import logging
import pprint
import random
import time
from collections import namedtuple
from dataclasses import dataclass
from typing import Optional

import numpy as np

from . import _evolvable_substrate
from ._functions import functions
from .cppn import CPPN
from .point import Point
from .config import Config


logger = logging.getLogger(__name__)


class ANN:
    @dataclass
    class Stats:
        axons: float = -1
        depth: np.uint = -1
        hidden: np.uint = -1
        edges: np.uint = -1
        density: float = -1
        iterations: np.uint = -1
        time: dict = dataclasses.field(
            default_factory=lambda: dict(build=-1, eval=-1))

        def dict(self): return dataclasses.asdict(self)

    def __init__(self):
        self._matrix: Optional[np.ndarray] = None
        self._inputs_idx, self._outputs_idx = {}, {}
        self._inputs, self._hidden, self._outputs = None, None, None
        self.NI, self.NH, self.NO = 0, 0, 0
        self._stats = ANN.Stats()
        self._act_function = functions[Config.activation_function]

    def reset(self):
        self._hidden[:] = 0

    def __repr__(self):
        return "Pure-Python ANN"

    def neuron_at(self, p: Point):
        raise NotImplementedError("Though I really should")

    def ibuffer(self): return self._inputs
    def obuffer(self): return self._outputs
    def buffers(self): return self.ibuffer(), self.obuffer()

    def __call__(self, ibuffer, obuffer, substeps: int = 1):
        start = _time()

        # print(ibuffer)
        assert ibuffer is self._inputs
        # print(self._inputs.shape, self._hidden.shape, self._matrix.shape)
        output = self._matrix.dot(np.concatenate(([1], self._inputs, self._hidden)))
        output = self._act_function(output)
        # print(output.shape)
        # print(output)
        self._hidden[:] = output[:self.NH]
        assert obuffer is self._outputs
        self._outputs[:] = output[self.NH:]

        self._stats.time['eval'] += _time_diff(start)

    def stats(self): return self._stats

    def empty(self, strict: bool = False):
        if strict:
            return self._stats.hidden == 0
        else:
            return self._stats.edges == 0

    def perceptron(self):
        return self.empty(strict=True) and not self.empty(strict=False)

    @classmethod
    def build(cls, inputs, outputs, genome):
        """
        Build an ANN

        :param inputs:
        :param outputs:
        :param genome:
        :return:
        """
        start = _time()

        cppn = CPPN(genome)
        ann = ANN()

        hidden, connections, iterations = (
            _evolvable_substrate.connect(cppn, inputs, outputs))

        ann.NI, ann.NH, ann.NO = len(inputs), len(hidden), len(outputs)

        ann._inputs = np.zeros(shape=(ann.NI,))
        ann._hidden = np.zeros(shape=(ann.NH,))
        ann._outputs = np.zeros(shape=(ann.NO,))

        ann._matrix = np.zeros(shape=(ann.NH+ann.NO, 1+ann.NI+ann.NH, ))

        for i, p in enumerate(inputs + hidden):
            ann._inputs_idx[p] = i+1

        for i, p in enumerate(hidden + outputs):
            ann._outputs_idx[p] = i
            ann._matrix[ann._outputs_idx[p], 0] = cppn(p, Point.null(),
                                                       CPPN.Output.BIAS)

        # print(f"[kgd-debug:{__name__}] hidden neurons: {pprint.pformat(hidden)}")

        for c in connections:
            ann._matrix[ann._outputs_idx[c.dst], ann._inputs_idx[c.src]] = c.weight

        ann._compute_stats(connections)
        ann._stats.iterations = iterations
        ann._stats.time['build'] = _time_diff(start)
        ann._stats.time['eval'] = 0
        return ann

    def _compute_stats(self, connections):
        self._stats.edges = len(connections)
        self._stats.axons = sum([(c.dst - c.src).length() for c in connections])

        self._stats.hidden = self.NH
        self._stats.density = self._stats.edges
        if self.NH == 0:
            self._stats.depth = 1
            self._stats.density /= self.NI * self.NO
        else:
            self._stats.depth = self._compute_depth(connections)
            self._stats.density /= (self.NI * self.NH + self.NH * self.NO)

    def _compute_depth(self, connections):
        return -1


def _time(): return time.perf_counter()
def _time_diff(start): return _time() - start
